### PR TITLE
Implemented deno.chmodSync for mac/linux os users.

### PR DESCRIPTION
**Pull request by srijanreddy98**
I have implemented chmodSync for mac/linux os.
Prior art:
Go: func Chmod(name string, mode FileMode) error

Node: fs.chmod(file, mode, callback){

Python: os.chmod(path, mode);

pub fn set_permissions<P: AsRef>(path: P, perm: Permissions) -> Result<()>

Permissions struct in Rust doesn't allow anything apart from -r--r--r-- or -rw-rw-rw

### CODE REVIEW

File: js/os_test.ts [hunk 1]
```diff
@@ -204,3 +204,22 @@ testPerm({ write: false }, function mkdDirSyncPerm() {
   assertEqual(err.name, "deno.PermissionDenied");
 });
 
+testPerm ({ write: true }, function chmodSync() {
+  const path = "/README.md";
+    const mode = "777";
+    deno.chmodSync(path, mode);
+});
```
Comment (ry) [inline js/os_test.ts]: It would be better to create a temp directory and use that for the chmod target.

File: js/os.ts [hunk 2]
```diff
@@ -336,3 +336,29 @@ export function writeFileSync(
...
+
+/**
+ * Changes the permissions of file or folder
+ *     import { chmodSync } from "deno";
+ *     const path = 'path';
+ *     const mode = '755'; // octal number in string
+ *
+ *     chmodSync(path, mode);
+ */
+export function chmodSync(path: string, mode: string): void {
```
Comment (ry) [inline js/os.ts]: I think we should use an integer for the `mode` - as Node does - https://nodejs.org/api/fs.html#fs_file_modes

File: src/handlers.rs [hunk 3]
```diff
@@ -673,3 +675,41 @@ fn handle_timer_clear(
...
+    Command::new("cmd")
+            .args(&["/C", "Yet to make for windows"])
+            .output()
+            .expect("failed to execute process")
+  } else {
+    Command::new("sh")
+            .arg("-c")
+            .arg(final_cmd.to_string())
+            .output()
+            .expect("failed to execute process")
```
Comment (ry) [inline src/handlers.rs]: Starting a new process to do `chmod` is much too much overhead. Please call the syscall directly.

**Comment by srijanreddy98** [comment]
I will start working on the suggested changes.

**Comment by kevinkassimo** [comment]
@srijanreddy98 Looks like you pushed `depot_tools_temp/` to the branch (639,880 new changes). Probably need to clean up

**Comment by srijanreddy98** [comment]
@kevinkassimo I removed them, but build fails at LINK ./deno

File: src/handlers.rs [hunk 4]
```diff
@@ -673,3 +678,36 @@ fn handle_timer_clear(
...
+    );
+    return Err(err.into());
+  }
+  let buf = CString::new(path).unwrap();
+  let _mode = CString::new(mode).unwrap();
+  let mut path_to_dynamic_lib = "";
+  if cfg!(target_os = "windows") {
+    path_to_dynamic_lib = "/home/srijan/Documents/chmod.dll"; // This will change to the path where deno will be present
+  } else {
+    path_to_dynamic_lib = "/home/srijan/Documents/chmod.so";  // This will change to the path where deno will be present
```
Comment (ry) [inline src/handlers.rs]: Hi Srijan - Thanks for looking into this but we shouldn't need any DLL loading to make this work. I think everything is in Rust's std library.

Comment (kevinkassimo) [reply src/handlers.rs]: @srijanreddy98 probably could try [std::fs::set_permissions](https://doc.rust-lang.org/std/fs/fn.set_permissions.html)

Comment (srijanreddy98) [reply src/handlers.rs]: @kevinkassimo I initially thought of using it but this method takes a permission struct which has only one method called set_readonly(x: bool). So you can only set permissions to -r--r--r-- or -rw-rw-rw-. So I was looking for other ways.

Comment (kevinkassimo) [reply src/handlers.rs]: @srijanreddy98 then probably [`PermissionExt`](https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html) is what you are looking for.
I believe [`OpenOptionsExt`](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html) could also be used/

**Comment by srijanreddy98** [comment]
@kevinkassimo  will it work for windows?

File: src/handlers.rs (hunk 4 above)
Comment (srijanreddy98) [reply src/handlers.rs]: @kevinkassimo will it work for windows?

**Comment by ry** [comment]
@piscisaureus How should chmod operate on windows?

File: src/handlers.rs [hunk 5]
```diff
@@ -670,3 +671,26 @@ fn handle_rename(d: *const DenoC, base: &msg::Base) -> Box<Op> {
...
+  let path = msg.path().unwrap();
+  let mode = msg.mode();
+  debug!("handle_chmod {} {}", path, mode);
+   if cfg!(target_family = "unix") {
+    Box::new(futures::future::result(|| -> OpResult {
+    let mut permissions = fs::metadata(path)?.permissions();
+    permissions.set_mode(mode);
+    let _result = fs::set_permissions(path, permissions)?;
+    Ok(None)
+  }()))
```
Comment (ry) [inline src/handlers.rs]: This code seems to be badly formatted. Please run ./tools/format.py

File: js/chmod_test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import { test, testPerm, assert, assertEqual } from "./test_util.ts";
+import * as deno from "deno";
+testPerm({ write: true }, function chmodSyncSuccess() {
+  const enc = new TextEncoder();
+  const data = enc.encode("Hello");
+  const filename = deno.makeTempDirSync() + "/test.txt";
+  deno.writeFileSync(filename, data, 0o666);
+  let fileInfo = deno.statSync(filename);
+  console.log(fileInfo.mode, 0o666); // This is printing 33206 438
```
Comment (srijanreddy98) [inline js/chmod_test.ts]: @ry deno.statSync(filename).mode is not working properly.

Checkout the line 802 in the build log [Here](https://travis-ci.com/denoland/deno/jobs/147768131). 
802: 33206 438. This is coming from the line console.log(fileInfo.mode, 0o666); on which I am commenting. 
I had initially opened an issue https://github.com/denoland/deno/issues/756. At that time it worked on travis. Now it is not working on travis also. Can you check if you are also having the same behaviour.

**Comment by ry** [comment]
@srijanreddy98 Sorry for letting this fall to the bottom - I was waiting on a discussion about windows support. I think simply having this being a no-op on windows is the right behavior (or at least acceptable for now). @kevinkassimo has cleaned it up in #1088. I will land there. 
Thank you

**Pull request closed without merging**